elastic/kibana
# [Fleet] Support for remote ES output

## Summary

Resolves https://github.com/elastic/kibana/issues/104986

Opening up for review, the feature flag is off for now, and the TODO items can come in follow up prs.

TODO:
- make service_token a secret field in output  - depends on https://github.com/elastic/kibana/issues/157458
- should link to remote elasticsearch docs in UI - depends on https://github.com/elastic/ingest-docs/issues/530
- remote es connection check and report on UI - depends on fleet-server to report unhealthy status if can't access the remote ES cluster
- enable feature flag when feature is ready

Added Remote ES output type, support to generate service token for `fleet-server-remote` account, support to create and edit remote es output.
Added validation to disallow making remote ES output as default for integration data.

## How to test locally?
Enable feature flag by adding this to `kibana.dev.yml`:
```
xpack.fleet.enableExperimental: ['remoteESOutput']
```
See e2e test instructions here: https://github.com/elastic/fleet-server/pull/3051

## Generate service token

Create remote service token API:
```
POST kbn:/api/fleet/service_tokens
{
  "remote": true
}

// kibana logs out
[2023-10-19T16:22:05.776+02:00][DEBUG][plugins.fleet] Creating service token for account elastic/fleet-server-remote
```

## Add/Edit output flyout:
Add output flyout:

<img width="675" alt="image" src="https://github.com/elastic/kibana/assets/90178898/dafc7d0e-05be-467f-871c-c4256fc833f6">

Edd output flyout:

<img width="660" alt="image" src="https://github.com/elastic/kibana/assets/90178898/0d58fcfb-8c22-4e27-8719-db86ecba2e8d">

Remote ES output not allowed to be set as integrations data output in agent policies, only as monitoring output:
<img width="690" alt="image" src="https://github.com/elastic/kibana/assets/90178898/675279cd-1c89-4069-9e07-e448aa796885">
<img width="683" alt="image" src="https://github.com/elastic/kibana/assets/90178898/6f67179d-b971-497f-9b04-3d3db5a42976">


Example API call to create/update output:
```
POST kbn:/api/fleet/outputs
{"name":"remote1","type":"remote-elasticsearch","hosts":["http://localhost:9200"],"is_default":false,"is_default_monitoring":false,"config_yaml":"","service_token":"token1","proxy_id":null}

PUT kbn:/api/fleet/outputs/39168010-6db8-11ee-9bf3-ed5492034535
{"name":"remote2","type":"remote-elasticsearch","hosts":["http://localhost:9200"],"is_default":false,"is_default_monitoring":false,"config_yaml":"","service_token":"token2","proxy_id":null}
```

### Checklist
